transactions/WFLY-17742_SupportTransactionsDuringGracefulShutdown.adoc

```diff
@@ -0,0 +1,122 @@
+
+=== Narayana
+
+Narayana will internally handle the lifecycle of transactions during suspension. From the point of view of the integrating party, Narayana should provide a blocking API to suspend itself and it should return control only when there are no transactions left to complete. Moreover, Narayana needs to provide a switch to suspend the creation of new transactions, which needs to be used only when no new transactions are needed.
+
+=== WildFly
+
+At the moment, WildFly’s graceful shutdown cannot be employed in cloud environments out of the box, especially when it comes to handling transactions during its suspension. In fact, as proved with WildFly’s Kubernetes Operator, before scaling down a pod hosting WildFly, all transactions must be completed.
+Even though WildFly’s graceful shutdown already gives the possibility to wait indefinitely, the Transactions subsystem does not really take advantage of this feature. Moreover, there is also an open question about whether the graceful shutdown should be processed synchronously or asynchronously. Currently, the list of subsystems registered with WildFly’s Suspend Controller is processed synchronously, which has its pros and cons.
+
+==== Modifications to WildFly’s Graceful Shutdown (wildfly-core)
+
+Following, modifications to WildFly’s graceful shutdown are discussed in more detail.
+
+*Sync or Async? ->* With regards to the open question presented in the overview, WildFly’s SuspendController synchronously calls pre-suspend and suspend on ServerActivity implementations. As a single timer is started to handle the overall graceful shutdown’s timeout, `ServerActivity` implementations (SAI) have to return very quickly from their pre-suspend/suspend routines to give enough time to other SAIs to process their pre-suspension/suspension. This means that timeout diminishes as SAIs process their suspension. This might result in prematurely cancelling the graceful shutdown for the last SAI in the queue as the overall timeout might have already elapsed. On the other hand, with a synchronous handling, WildFly suspends all registered SAIs one by one, which means that the next SAI does not receive the signal to pre-suspend/suspend until the current SAI has returned the call to pre-suspend/suspend. With an asynchronous suspension, this will no longer be the case and all SAIs will initiate their pre-suspension/suspension concurrently. As a consequence, SAIs should coordinate better their suspensions instead of relying on the intrinsic isolation of the synchronous approach.
+*Proposed modifications.* Introducing an asynchronous graceful shutdown would enable all SAIs to receive a pre-suspend/suspend signal concurrently. In this way, the timeout will have the same duration for all SAIs.
```

Comment:
> If we go with this, we need to add hard requirements to analyze the existing ServerActivities, identify any ordering issues and specify how the necessary coordination will happen.
> 
> I think doing that is a worthwhile exercise, as 'relying on the intrinsic isolation of the synchronous approach' is really fragile IMO. The dependency relationships between services are not likely to have been designed with this suspend ordering at the front of mind, and if it was it's also likely that it wasn't clearly written down and could be easily broken by maintainers. Plus there's the timeout issue you point out.

Thank you for your input, Brian. I share the view that a more thorough analysis of the `ServerActivities` area is necessary. However, I suggest we exclude that analysis from the current proposal. In fact, I believe the analysis and the associated changes fall beyond the scope of the design presented here. Instead, I propose initiating a separate proposal (in the `core` folder of this repo) solely focused on altering how WildFly’s `SuspendController` invokes `ServerActivity` implementations' `pre-suspend` and `suspend` hooks. If you agree, I propose retaining only the "FIFO or LIFO" paragraph in the "Modifications to WildFly’s Graceful Shutdown (wildfly-core)" section. The "Sync or Async" paragraph could then be moved to a dedicated proposal within the core section. WDYT?

Reply:
+1. I think starting with LIFO is the way to go.

Reply:
I should have some time now to dig into the activities though, so I created https://docs.google.com/document/d/11RfZSzOCqqFbinckIRnVkuyF84HpIAF8zz1CG6k6Q_M/edit as a working document. That's more key for the async approach, but it may be useful in some way for the LIFO approach (i.e. by clarifying the order in which SAs are getting registered.)

Reply:
Hi @bstansberry! I modified the text of the proposal. I hope the new wording can resolve this last conversation. Thanks for your help with this :-)